### PR TITLE
Feat2: refactor: consolidate mint/burn operations into unified dispatch function

### PR DESCRIPTION
**Pull request by guil-lambert**
## Summary
Major refactoring of the PanopticPool's option management system, consolidating multiple mint/burn functions into a single `dispatch` function that handles all position operations. Additionally merges liquidation, force exercise, and premium settlement into a unified `dispatchFrom` function.

### Consolidated Entry Points
**Before:** Separate functions for each operation:
- `mintOptions()` - mint a single position
- `burnOptions()` - burn a single position  
- `burnOptions()` (overloaded) - burn multiple positions
- `liquidate()` - liquidate insolvent accounts
- `forceExercise()` - force exercise OTM positions
- `settleLongPremium()` - settle premium for long legs

**After:** Two unified dispatch functions:
- `dispatch()` - handles both mints and burns based on position state
- `dispatchFrom()` - handles liquidations, force exercises, and premium settlements based on solvency state

#### `dispatch()` Function
- Iterates through `positionIdList`
- For each tokenId:
  - If position doesn't exist → mint
  - If position exists → burn
- Single solvency check at the end with `finalPositionIdList`

#### `dispatchFrom()` Function  
- Determines operation type based on solvency:
  - If insolvent at all ticks → liquidation (path = 2)
  - If solvent at all ticks → settle premium (path = 0) or force exercise (path = 1)
- Validates final state for both caller and callee

### Internal Changes
- `_mintOptions()` simplified - no longer handles position validation or solvency checks
- `_burnOptions()` now takes `positionSize` as parameter
- `_checkSolvencyAtTicks()` returns count of solvent ticks instead of reverting
- `_validatePositionList()` simplified - removed offset parameter
- Moved `_settleLongPremium()`, `_forceExercise()`, and `_liquidate()` to internal functions

## Benefits

1. **Reduced Complexity**: Single entry point for all position operations
2. **Atomic Operations**: Can mint and burn in a single transaction
3. **Gas Optimization**: Fewer external calls and duplicate checks
4. **Improved UX**: Simpler interface for integrators
5. **Better Error Handling**: Unified validation and solvency checking
6.  **Bytecode Optimization**: This refactoring achieves a 2.35kb reduction in contract bytecode through consolidation of redundant code paths and removal of duplicate validation logic. This creates room for future enhancements without approaching size limits

## Breaking Changes

⚠️ **This is a major breaking change** that affects all integrations:

### Removed Functions
- `mintOptions(TokenId[], uint128, uint64, int24, int24, bool)`
- `burnOptions(TokenId, TokenId[], int24, int24, bool)`  
- `burnOptions(TokenId[], TokenId[], int24, int24, bool)`
- `liquidate(TokenId[], address, TokenId[])`
- `forceExercise(address, TokenId, TokenId[], TokenId[], LeftRightUnsigned)`
- `settleLongPremium(TokenId[], address, uint256, bool)`

### New Functions
- `dispatch(TokenId[], TokenId[], uint128[], uint64[], int24, int24, bool)`
- `dispatchFrom(TokenId[], address, TokenId[], TokenId[], LeftRightUnsigned)`

### Migration Guide
Integrators will need to:
1. Update all mint calls to use `dispatch()` with appropriate parameters
2. Update all burn calls to use `dispatch()` with zero position sizes
3. Update liquidation/exercise calls to use `dispatchFrom()`
4. Adjust for new parameter structures (arrays instead of single values)

## Testing
- [ ] All existing tests updated for new interface
- [ ] Gas benchmarks comparing old vs new implementation
- [ ] Integration tests for mixed mint/burn operations
- [ ] Edge cases for solvency determination paths

## Security Considerations
- Solvency checks moved to end of operations - ensure no intermediate unsafe states
- Path determination logic in `dispatchFrom()` must be carefully audited
- Array length validations critical for preventing mismatched operations

## Summary by CodeRabbit

- New Features
  - Unified multi-asset dispatch/transfer entry points and ERC‑1155 single-transfer support.
  - Option to treat accrued premia as collateral, affecting withdraw/settle behavior.

- Changes
  - Public interfaces extended to accept position lists and premia-as-collateral flag.
  - Oracle-based TWAP adopted; stale-TWAP error renamed to “StaleOracle.”
  - Expanded liquidation, force-exercise, and solvency checks across multi-tick paths.

- Refactor
  - Internal mint/burn/settlement flows and math/position helpers reorganized for the new semantics.

- Tests
  - Test suite updated and new scenarios added to exercise token-substitution and settlement paths.

### CODE REVIEW

File: contracts/libraries/PanopticMath.sol [hunk 1]
```diff
@@ -128,10 +128,17 @@ library PanopticMath {
...
-            : uint8(existingHash >> 248) - tokenId.countLegs();
+
+        uint256 newLegCount;
+        if (addFlag) {
+            newLegCount = uint8(existingHash >> 248) + uint8(positionLegs);
+        } else {
+            unchecked {
+                newLegCount = (existingHash >> 248) - positionLegs;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (guil-lambert) [reply contracts/libraries/PanopticMath.sol:141]: This is the intended behavior

File: contracts/PanopticPool.sol [hunk 2]
```diff
@@ -468,69 +494,69 @@ contract PanopticPool is ERC1155Holder, Multicall {
...
+            finalPositionIdList,
+            BP_DECREASE_BUFFER,
+            usePremiaAsCollateral
         );
 
-        uint256 medianData = _validateSolvency(msg.sender, newPositionIdList, NO_BUFFER);
-
         // Update `s_miniMedian` with a new observation if the last observation is old enough (returned medianData is nonzero)
         if (medianData != 0) s_miniMedian = medianData;
     }
```
[the remark this answers is not in the document]
Comment (guil-lambert) [reply contracts/PanopticPool.sol:562]: Yes, the leg increment was needed, thanks!

What could happen if there were no length validations except failing/reverting?

File: contracts/PanopticPool.sol [hunk 3]
```diff
@@ -1401,31 +1547,31 @@
...
+        uint256 removedLiquidity;
         (totalLiquidity, netLiquidity, removedLiquidity) = _getLiquidities(tokenId, leg);
 
         // compute and return effective liquidity. Return if short=net=0, which is closing short position
         if (netLiquidity == 0 && removedLiquidity == 0) return totalLiquidity;
 
         uint256 effectiveLiquidityFactorX32;
         unchecked {
-            effectiveLiquidityFactorX32 = (uint256(removedLiquidity) * 2 ** 32) / netLiquidity;
+            effectiveLiquidityFactorX32 = (removedLiquidity * 2 ** 32) / netLiquidity;
```
[the remark this answers is not in the document]
Comment (guil-lambert) [reply contracts/PanopticPool.sol:1559]: netLiquidity can never be zero
